plutus-ir/src/Language/PlutusIR/Analysis/Free.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+-- | Functions for computing the free (type-)variables of a PIR term or type.
+module Language.PlutusIR.Analysis.Free (fTerm, fType, fVarDecl, fBinding) where
+
+import           Language.PlutusIR
+
+import qualified Language.PlutusCore       as PLC
+import qualified Language.PlutusCore.Name  as PLC
+
+import           Language.PlutusCore.Subst (ftvTy)
+
+import           Control.Lens
+
+import qualified Data.Set                  as S
+
+{- Note: [PIR Free variables]
+
+The functions of this module behave similar to those found in 'language-plutus-core/src/Language/PlutusCore/Subst.hs',
+with the difference:
+
+- extended to support the plutus-ir (letrec and letnonrec)
+- place the free variables and free type variables in the same set of uniques.
+
+The functions do not require that the input term be prior 'renamed'/'uniqueified'.
+-}
+
+
+{- Note [Free variables and letnonrec linear scope]
+
+Let (nonrec) is linearly scoped, as shown by the example:
+
+let b1 = rhs1;
+    b2 = rhs2  (b1 is visible in rhs2);
+in ...
+
+is the same as as let b1 = rhs in (let b2 = rhs2 in ... )
+
+And is different than letrec where all identifiers are visible/scoped:
+
+let b1 = rhs1 (b2 is visible in rhs1);
+    b2 = rhs2  (b1 is visible in rhs2);
+in ...
+
+
+In this regard, extra care has been taken when counting the free variables of a letnorec vs a letrec, as
+shown by the example:
+
+(let nonrec { x = y
+              y = 3 }
+) ===> y is free in x
+
+VS
+
+(let rec { x = y
+           y = 3 }
+) ===> y is not free in x
+
+-}
+
+
+-- | Given a term, it returns a set of all the FREE variables inside that term (i.e. not declared/scoped inside the term).
+--
+-- It does not require the input term to be prior 'PLC.rename'd.
+-- TODO: refactor using recursion-schemes (or usingtermSubterms, termSubtypes can be used and lens + fold), see 'language-plutus-core/src/Language/PlutusCore/Subst.hs'
+-- See Note: [PIR Free variables]
+fTerm :: forall name tyname a.
+  (Ord (tyname a), PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+  => Term tyname name a -> S.Set PLC.Unique
+fTerm = \case
+  Var _ x -> S.singleton $ x ^. PLC.unique . coerced
+
+  Let _ NonRec bs tIn -> fLet bs tIn NonRec
+  Let _ Rec bs tIn -> fLet bs tIn Rec
+
+  TyAbs _ n _ t -> S.delete (n ^. PLC.unique . coerced) $ fTerm t
+  LamAbs _ n ty t -> fType ty
+                     <>
+                     S.delete (n ^. PLC.unique . coerced) (fTerm t)
+  Apply _ t1 t2 -> fTerm t1 <> fTerm t2
+  TyInst _ t ty -> fTerm t <> fType ty
+  IWrap _ ty1 ty2 t -> mconcat [fType ty1, fType ty2, fTerm t]
+  Unwrap _ t -> fTerm t
+  _ -> S.empty
+
+ where
+  fLet :: [Binding tyname name a] -> Term tyname name a -> Recursivity -> S.Set PLC.Unique
+  fLet bs tIn =
+    let fIn = fTerm tIn S.\\ foldMap bindingIds bs -- the free variables of termIn (scoped by all the let id's)
```

Comment:
We have two scopes: type-level names and term-level names. `bindingIds` return the ids of all names in both the scopes. And here we remove all ids referenced in `Binding`s from free variables of a term. So what if one of the `Binding`s binds a type variable with the `0` unique and `tIn` has a free variable with the same unique? Despite sharing the unique, those are two distinct names, but they will be equated here.

If this behaviour is indeed what you want, it's worth adding a note explaining why conflating scopes is fine here. We have one such note in `Language.PlutusCore.Generators.AST`:

```haskell
{- Note [ScopeHandling]
We intentionally do not distinguish between the type-level and term-level scopes to ensure that
all the machineries handle variables with same uniques from distinct scopes correctly.
-}
```

Reply:
I think this is a bug indeed. If we need Free to work w/o renaming in place, we need to fix this by having two separate scopes (term & type)

Reply:
> I think this is a bug indeed. If we need Free to work w/o renaming in place, we need to fix this by having two separate scopes (term & type)

The current IR needs a fix to help here, because it doesn't account for free variables:

```haskell
instance PLC.HasUniques (Term tyname name uni ann) => PLC.Rename (Term tyname name uni ann) where
    -- TODO: the Plutus Core codebase uses marking in order to prevent clashing with existing
    -- free variables. Should we do the same here?
    rename = PLC.runRenameT . renameTermM
```

so the renamer might rename a type binding into the same unique that a term-level free variable has. In Plutus Core we have `Language.PlutusCore.Mark` that allows to bump the current fresh unique to the one that is greater than all uniques of a type/term. See `Language.PlutusCore.Rename` for how we use marking:

```haskell
{- Note [Marking]
We use functions from the @markNonFresh*@ family in order to ensure that bound variables never get
renamed to free ones. This means types/terms/etc are processed twice: once by a @markNonFresh*@
function and once for the actual renaming.
-}

instance HasUniques (Term tyname name uni ann) => Rename (Term tyname name uni ann) where
    -- See Note [Marking].
    rename = through markNonFreshTerm >=> runRenameT . renameTermM
```

I think both relying on the renamer or tracking the two scopes separately is a pain in the ass, because the former can easily lead to unexpected behaviour (happened way too many times for me, it's really easy to overlook something when it comes to unique-based name handling) and the latter can be too convoluted.

I think we should fix the IR renamer and require the term to be renamed just to get this PR going. We can consider tracking scopes later.